From 7d329b7ff8b92dee0a4a2f014b33c2d2f2aa9834 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Fri, 1 Oct 2021 16:41:48 +0200
Subject: [PATCH 145/841] feat(stm32mp2): enable DDR PHY driver

Call init sequence in DDR driver. Add all related variables, include
directories and source files in the platform makefile.
Increase BL2 size to support this new driver.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git fbf521fca63d075c5be947996fff7340d918166d feat(stm32mp2): enable DDR PHY driver]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I2fbdb5ffc3929cc975c1b28eb0d2ce9eb320daac
---
 drivers/st/ddr/stm32mp2_ddr.c | 12 ++++++++++--
 plat/st/stm32mp2/platform.mk  | 25 +++++++++++++++++++++++++
 2 files changed, 35 insertions(+), 2 deletions(-)

diff --git a/drivers/st/ddr/stm32mp2_ddr.c b/drivers/st/ddr/stm32mp2_ddr.c
index d9682e8b2..341d2b7f1 100644
--- a/drivers/st/ddr/stm32mp2_ddr.c
+++ b/drivers/st/ddr/stm32mp2_ddr.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2021, STMicroelectronics - All Rights Reserved
+ * Copyright (C) 2021-2022, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -13,8 +13,12 @@
 #include <drivers/st/stm32mp2_ddr_regs.h>
 #include <lib/mmio.h>
 
+#include <ddrphy_phyinit.h>
+
 #include <platform_def.h>
 
+#define SKIP_TRAINING_PARAM	false
+
 #define DDRDBG_FRAC_PLL_LOCK	U(0x10)
 
 #define DDRCTL_REG(x, y)					\
@@ -386,7 +390,11 @@ void stm32mp2_ddr_init(struct stm32mp_ddr_priv *priv,
 
 	disable_refresh(priv->ctl);
 
-	/* TODO execute DDR PHY init sequence */
+	ret = ddrphy_phyinit_sequence(SKIP_TRAINING_PARAM);
+	if (ret != 0) {
+		ERROR("DDR PHY init: Error %d\n", ret);
+		panic();
+	}
 
 	activate_controller(priv->ctl);
 
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index 8d18f639a..186005470 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -114,6 +114,8 @@ TF_CFLAGS			+=	-mbranch-protection=none
 
 # Include paths and source files
 PLAT_INCLUDES			+=	-Iplat/st/stm32mp2/include/
+PLAT_INCLUDES			+=	-Idrivers/st/ddr/phy/phyinit/include/
+PLAT_INCLUDES			+=	-Idrivers/st/ddr/phy/firmware/include/
 
 PLAT_BL_COMMON_SOURCES		+=	lib/cpus/${ARCH}/cortex_a35.S
 PLAT_BL_COMMON_SOURCES		+=	drivers/st/uart/${ARCH}/stm32_console.S
@@ -158,6 +160,29 @@ BL2_SOURCES			+=	drivers/st/ddr/stm32mp2_ddr.c				\
 					drivers/st/ddr/stm32mp2_ddr_helpers.c			\
 					drivers/st/ddr/stm32mp2_ram.c
 
+BL2_SOURCES			+=	drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_c_initphyconfig.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_calcmb.c					\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_globals.c					\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_initstruct.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_isdbytedisabled.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_loadpieprodcode.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_mapdrvstren.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_softsetmb.c				\
+					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_customposttrain.c	\
+					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_custompretrain.c
+
+BL2_SOURCES			+=	drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_d_loadimem.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_f_loaddmem.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_g_execfw.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_h_readmsgblock.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_storemsgblk.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_writeoutmem.c				\
+					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_h_readmsgblock.c 	\
+					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_g_waitfwdone.c
+
 # BL31 sources
 BL31_SOURCES			+=	${FDT_WRAPPERS_SOURCES}
 
-- 
2.47.2

